edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubernetesServiceBuilder.cs

```diff
@@ -0,0 +1,163 @@
+                    var annotations = new Dictionary<string, string>();
+                    foreach (KeyValuePair<string, string> label in moduleWithDockerConfig.Config.CreateOptions?.Labels)
+                    {
+                        annotations.Add(KubeUtils.SanitizeAnnotationKey(label.Key), label.Value);
+                    }
+
+                    serviceAnnotations = Option.Some(annotations);
+                }
+
+                // Handle ExposedPorts entries
+                if (moduleWithDockerConfig.Config?.CreateOptions?.ExposedPorts != null)
+                {
+                    // Entries in the Exposed Port list just tell Docker that this container wants to listen on that port.
+                    // We interpret this as a "ClusterIP" service type listening on that exposed port, backed by this module.
+                    // Users of this Module's exposed port should be able to find the service by connecting to "<module name>:<port>"
+                    this.GetExposedPorts(moduleWithDockerConfig.Config.CreateOptions.ExposedPorts)
+                        .ForEach(
+                            exposedList =>
+                                exposedList.ForEach((item) => portList.Add(new V1ServicePort(item.Port, name: $"ExposedPort-{item.Port}-{item.Protocol.ToLower()}", protocol: item.Protocol))));
+                }
+
+                // Handle HostConfig PortBindings entries
+                if (moduleWithDockerConfig.Config?.CreateOptions?.HostConfig?.PortBindings != null)
+                {
+                    foreach (KeyValuePair<string, IList<DockerModels.PortBinding>> portBinding in moduleWithDockerConfig.Config?.CreateOptions?.HostConfig?.PortBindings)
+                    {
+                        string[] portProtocol = portBinding.Key.Split('/');
+                        if (portProtocol.Length == 2)
+                        {
+                            if (int.TryParse(portProtocol[0], out int port) && ProtocolExtensions.TryValidateProtocol(portProtocol[1], out string protocol))
+                            {
+                                // Entries in Docker portMap wants to expose a port on the host (hostPort) and map it to the container's port (port)
+                                // We interpret that as the pod wants the cluster to expose a port on a public IP (hostPort), and target it to the container's port (port)
+                                foreach (DockerModels.PortBinding hostBinding in portBinding.Value)
+                                {
+                                    if (int.TryParse(hostBinding.HostPort, out int hostPort))
+                                    {
+                                        // If a port entry contains the same "port", then remove it and replace with a new ServicePort that contains a target.
+                                        var duplicate = portList.SingleOrDefault(a => a.Port == hostPort);
```

Comment:
Probably we can use `HashSet` with custom equality comparer for storing ports. In this case, we don't need to manually iterate across list and ensure unique elements

Reply:
We could! I want this PR to simply be a refactor, then I will add back my UTs, and then make improvements a-la TDD style :)